pageserver/src/virtual_file.rs

```diff
@@ -566,12 +580,19 @@ impl VirtualFile {
 impl Drop for VirtualFile {
     /// If a VirtualFile is dropped, close the underlying file if it was open.
     fn drop(&mut self) {
-        let handle = self.handle.get_mut().unwrap();
-
-        // We could check with a read-lock first, to avoid waiting on an
-        // unrelated I/O.
+        let handle = self.handle.get_mut();
+
+        // We don't have async drop so we cannot wait for the lock here.
+        // Instead, do a best-effort attempt at closing the underlying
+        // file descriptor by using `try_write`.
+        // This best-effort attempt should be quite good though
+        // as we have `&mut self` access. In other words, if the slot
+        // is still occupied by our file, we should be the only ones
+        // accessing it (and if it has been reassigned since, we don't
+        // need to bother with dropping anyways).
         let slot = &get_open_files().slots[handle.index];
-        let mut slot_guard = slot.inner.write().unwrap();
+        let Ok(mut slot_guard) = slot.inner.try_write() else { return };
+
```

Comment:
Hmm, so if we don't get the lock, then the fd is kept open until it is evicted from the open files list. That's OK from a correctness point of view. I'm a little worried, though, that we might hold onto a large file that was deleted, so that the disk space cannot be reused.

Reply:
Some alternatives:
- retry a few times, with a small delay. This is expected to be very rare, so that should be fine
- add a little global queue of dropped slots. Check that queue periodically and close all the underlying files
- just accept this as it is. Add an info! message for when it happens, so that we can keep an eye on it if it becomes a problem

I'm ok with any of those

Reply:
> retry a few times, with a small delay. This is expected to be very rare, so that should be fine

Personally I'd go with this.

As explained in the PR description, the only other component that can access the lock and cause the `try_lock` to fail is some other thread running the clock algorithm to get a free slot and is over that specific slot right in the moment where we drop the `VirtualFile`. This is unlikely, but gets more likely the more pressure is on the cache. In high-pressure situations however, leakages are being cleaned up quickly.

Reply:
>For the rare else branch, I'd just spawn a task, do the write().await there.

Ah yes, that's a good idea, simple and robust